pkg/coderef/api/code_reference.go

```diff
@@ -0,0 +1,618 @@
+	}
+	if req.FileExtension != "" {
+		whereParts = append(whereParts, mysql.NewFilter("file_path", "LIKE", "%"+req.FileExtension))
+	}
+	orders := []*mysql.Order{mysql.NewOrder("id", mysql.OrderDirectionAsc)}
+	switch req.OrderBy {
+	case proto.ListCodeReferencesRequest_CREATED_AT:
+		orders = []*mysql.Order{mysql.NewOrder("created_at", s.toMySQLOrderDirection(req.OrderDirection))}
+	case proto.ListCodeReferencesRequest_UPDATED_AT:
+		orders = []*mysql.Order{mysql.NewOrder("updated_at", s.toMySQLOrderDirection(req.OrderDirection))}
+	}
+	limit := int(req.PageSize)
+	cursor := 0
+	if req.Cursor != "" {
+		c, err := strconv.Atoi(req.Cursor)
+		if err != nil {
+			dt, err := statusInvalidCursor.WithDetails(&errdetails.LocalizedMessage{
+				Locale:  localizer.GetLocale(),
+				Message: localizer.MustLocalize(locale.InvalidArgumentError),
+			})
+			if err != nil {
+				return nil, statusInternal.Err()
+			}
+			return nil, dt.Err()
+		}
+		cursor = c
+	}
+	codeRefStorage := v2.NewCodeReferenceStorage(s.mysqlClient)
+	codeRefs, nextCursor, totalCount, err := codeRefStorage.ListCodeReferences(
+		ctx,
+		whereParts,
+		orders,
+		limit,
+		cursor,
+	)
+	if err != nil {
+		s.logger.Error(
+			"Failed to list code references",
+			log.FieldsFromImcomingContext(ctx).AddFields(
+				zap.Error(err),
+				zap.String("environmentId", req.EnvironmentId),
+			)...,
+		)
+		dt, err := statusInternal.WithDetails(&errdetails.LocalizedMessage{
+			Locale:  localizer.GetLocale(),
+			Message: localizer.MustLocalize(locale.InternalServerError),
+		})
+		if err != nil {
+			return nil, statusInternal.Err()
+		}
+		return nil, dt.Err()
+	}
+	protoRefs := make([]*proto.CodeReference, 0, len(codeRefs))
+	for _, ref := range codeRefs {
+		ref.CodeReference.SourceUrl = generateSourceURL(&ref.CodeReference)
+		protoRefs = append(protoRefs, &ref.CodeReference)
+	}
+	return &proto.ListCodeReferencesResponse{
+		CodeReferences: protoRefs,
+		Cursor:         strconv.Itoa(nextCursor),
+		TotalCount:     totalCount,
+	}, nil
+}
+
+func (s *CodeReferenceService) CreateCodeReference(
+	ctx context.Context,
+	req *proto.CreateCodeReferenceRequest,
+) (*proto.CreateCodeReferenceResponse, error) {
+	localizer := locale.NewLocalizer(ctx)
+	editor, err := s.checkEnvironmentRole(
+		ctx,
+		accountproto.AccountV2_Role_Environment_EDITOR,
+		req.EnvironmentId,
+		localizer,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if err := validateCreateCodeReferenceRequest(req, localizer); err != nil {
+		return nil, err
+	}
+	id, err := uuid.NewUUID()
+	if err != nil {
+		s.logger.Error(
+			"Failed to generate uuid",
+			log.FieldsFromImcomingContext(ctx).AddFields(
+				zap.Error(err),
+			)...,
+		)
+		dt, err := statusInternal.WithDetails(&errdetails.LocalizedMessage{
+			Locale:  localizer.GetLocale(),
+			Message: localizer.MustLocalize(locale.InternalServerError),
+		})
+		if err != nil {
+			return nil, statusInternal.Err()
+		}
+		return nil, dt.Err()
+	}
```

Comment:
Please move the UUID logic to the `domain.NewCodeReference` as we do for the other entities.